### PR TITLE
Improve environment management

### PR DESCRIPTION
**Pull request by sdispater**
## Pull Request Check List

- [x] Added **tests** for changed code.
- [ ] Updated **documentation** for changed code.

This PR should fix or at least alleviate issues that users face when trying to setup a project with a python requirement that is not compatible with the currently used Python version (see #536).

The environment manager will now try to find a compatible Python version in the current path and use/activate it accordingly. Note that it will still fail if none can be found but with, hopefully, a better error message with some pointers as to how the user can solve the problem.

### CODE REVIEW

**Comment by peter-schmidbauer** [comment]
This seems to fix #1417 and makes #1469 obsolete.

**Comment by KelSolaar** [comment]
That's great thanks @sdispater !

File: poetry/utils/env.py [hunk 1]
```diff
@@ -449,9 +473,84 @@ def create_venv(
...
+
+            io.write_line(
+                "<warning>The currently activated Python version {} "
+                "is not supported by the project ({}).\n"
+                "Trying to find and use a compatible version.</warning> ".format(
+                    python_patch, self._poetry.package.python_versions
                 )
             )
 
+            for python_to_try in reversed(
```
Comment (Congee) [inline poetry/utils/env.py]: why not sorted(arg, arg, reverse=True)

**Comment by maggyero** [comment]
Hi @sdispater I have just tested your PR #1477 with the latest Poetry (1.0.0b5) on MacOS Catalina (10.15.1). (Sorry for not providing any feedback earlier.) Poetry now correctly creates virtual environments based on the requirements of `tool.poetry.dependencies.python` in pyproject.toml! However it does not follow the specifications in my initial post (https://github.com/sdispater/poetry/issues/1386#issue-495076159, section "[Edit] Specification for the second point").

Consequently, with the value `">=2.7"` for `tool.poetry.dependencies.python` in pyproject.toml, I get two different Python versions installed, _according to the Python running the `poetry` script_ (note: on my system, `python` and `python2` points to Python 2.7 and `python3` on Python 3.7):

- Python 2.7 is installed if `poetry` is run with Python 2.7:

```
[maggyero@MacBookPro~/Desktop/foo]$ python ~/.poetry/bin/poetry install 
Creating virtualenv foo-1XhTtlMR-py2.7 in /Users/maggyero/Library/Caches/pypoetry/virtualenvs
Installing dependencies from lock file
Warning: The lock file is not up to date with the latest changes in pyproject.toml. You may be getting outdated dependencies. Run update to update them.


Package operations: 1 install, 0 updates, 0 removals

  - Installing pika (1.1.0)
  - Installing foo (0.1.0)
```

- Python 3.7 is installed if `poetry` is run with Python 3.7:

```
[maggyero@MacBookPro~/Desktop/foo]$ python3 ~/.poetry/bin/poetry install   
Creating virtualenv foo-1XhTtlMR-py3.7 in /Users/maggyero/Library/Caches/pypoetry/virtualenvs
Installing dependencies from lock file
Warning: The lock file is not up to date with the latest changes in pyproject.toml. You may be getting outdated dependencies. Run update to update them.


Package operations: 1 install, 0 updates, 0 removals

  - Installing pika (1.1.0)
  - Installing foo (0.1.0)
```

Both installed Python respect the `">=2.7"` requirements but I would expect Poetry to always install the same Python: the highest version available (Python 3.7 in this case), _regardless of the Python running `poetry`_.

**Pull request merged**